code/datums/craft/step.dm

```diff
@@ -67,34 +79,43 @@
 		if(!istype(I, reqed_type))
 			user << "Wrong item!"
 			return
-		if(req_amount && ispath(reqed_type, /obj/item/stack))
+		if (!is_valid_to_consume(I, user))
+			user << "That item can't be used for crafting!"
+			return
+
+		if(req_amount && istype(I, /obj/item/stack))
 			var/obj/item/stack/S = I
-			if(S.amount < req_amount)
-				user << "Not enought items in [I]"
+			if(S.get_amount() < req_amount)
+				user << "Not enough items in [I]"
 				return
+
+
 		if(target)
-			annonce_action(start_msg, user, I, target)
+			announce_action(start_msg, user, I, target)
 		if(!do_after(user, time, target || user))
 			return
 	else if(reqed_quality)
 		if(!I.get_tool_quality(reqed_quality))
-			user << "Wrong item!"
 			return
 		if(target)
-			annonce_action(start_msg, user, I, target)
+			announce_action(start_msg, user, I, target)
 		if(!I.use_tool(user, target || user, time, reqed_quality))
 			return
+	else
+		if(!do_after(user, time, target || user))
+			return
 
 	if(req_amount)
 		if(istype(I, /obj/item/stack))
 			var/obj/item/stack/S = I
 			if(!S.use(req_amount))
+				user << SPAN_WARNING("Not enough items in [S]. It has [S.get_amount()] units and we need [req_amount]")
 				return FALSE
-		else
+		else if (reqed_type) //No deleting tools
```

Comment:
You wanna make it impossible to consume items specified by type?

Reply:
no, that check makes it ONLY possible to consume items specified by type.

Meaning that, if we got here with a reqed quality, it will be false, and execution won't get in there to the qdel call
so it makes it impossible to consume your tools

Reply:
But that already checked in `is_valid_to_consume(...)`.
For things like bots items specified with quality can be useable and must be able to consume.
And that unobvious behavior if you can describe a step requiring a given quality and mark it as comsumeable. Furthemore, description will say, what item will be consumed, but it will not.

Reply:
I don't understand what you're saying here. 
The system does not support consuming tools. Specified type items are consumed. Required quality tools are used but not consumed. You can't select and consume something based on its qualities.

There's no existing use for that behaviour and removing this check would be a very messy and half-assed way to add it. If you want that behaviour we can code it seperately. But for now, only items specified by required type should be possible to consume

Reply:
You can define step as
`/obj/item/weapon/smth` or `QUALITY_SCREWDRIVING` - to "use item"
On the other hand you can define it with list:
`list(/obj/item/weapon/smth, amount)` or `list(QUALITY_SCREWDRIVING, amount)` to "consume item"

Reply:
Nooooo, the amount field for required qualities specifies the required level of that quality

We could add support for consuming tools in future, the crafting system is far from done. But it's beyond the scope of this PR. Afaik there is no current application for consuming your tools, and no reason to code that

Reply:
But qualities not working that way. That seems like they should, but right now they only affect times and probs, not restrict anything.
Any way that's not right plase for such check. You should catch it in step parsing and drop `consume amount` value then step will have correct description and so on.

Reply:
I feel like we're completely misunderstanding each other. Let me explain how i see this.
When a quality and an amount are specified in crafting, it means you need to use a tool with that amount of that qualitt. Eg QUALITY_BOLT_TURNING, 30 requires a wrench, normal tools have 30
The tool is used in the recipe but not consumed. It remains in your hand after that crafting step is done

The req_amount field is used for the required level of the quality. In the above example, that is 30.  It does not mean consume 30 wrenches, it means you need one wrench with a bolt turning quality of 30 or more
there is absolutely no support for eating tools. 

Tool qualities do not currently affect success rates or worktimes of crafting, thereare no provisions for that

For the second point;
It is a fine place for the check. If the check were removed, then tools would be consumed. IF thecks req amount - which is going to be true for tools because they have a numerical requirement for their quality
This works.

Tools will not be dropped, that's not what tools are for

Now closing arguments;

If you want to see crafting recipes which CONSUME the tools used in them, it is generally easy to do that by setting the tool as a required item instead of a required quality.

If you really want to see support for recipes which consume a tool selected by its quality, that is a completely seperate feature and it is beyond the scope of this PR

This is not the final crafting PR, it can and will be adjusted more in future. The purpose of this PR, as i said at the start, is to make crafting work for robots. This argument over tool consumption is pointless. If you want the ability to consume tools, add it. I am not opposing it. It's just not part of this work